src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -483,10 +483,11 @@ async fn optimize_merge_into(mut opt_ctx: OptimizerContext, s_expr: SExpr) -> Re
     let distributed = !join_s_expr.has_merge_exchange();
     if opt_ctx.enable_distributed_optimization && distributed {
         let merge_optimizer = MergeOptimizer::create();
-        // now always use shuffle
+        // If inner join && !join_order_changed, source will broadcast,
+        // others use shuffle.
         if matches!(
             join_op.join_type,
-            JoinType::Left | JoinType::LeftAnti | JoinType::Inner
+            JoinType::Left | JoinType::LeftAnti | JoinType::Inner if join_order_changed
```

Comment:
Why need to add the condition, can we directly use the cbo's distribution result?

Reply:
If the target is used as the build side, the CBO will make the target to broadcast, which will cause deduplication checks to occur in the main node.